### DISCUSSION TITLE
Wrong Average and Sum

### DISCUSSION THREAD

**Discussion by triposat** [data-forge/data-forge-ts#159] [General]
My **CSV** File:

<img width="423" alt="Screenshot 2023-03-08 213248" src="https://user-images.githubusercontent.com/69134468/223765560-d16314d2-9f3e-4c46-87f7-56883965e6cd.png">

**Code:**

<img width="497" alt="Screenshot 2023-03-08 213451" src="https://user-images.githubusercontent.com/69134468/223765534-c718ff15-795c-4426-ac04-9443c49e4d2f.png">

**Output:**

<img width="292" alt="Screenshot 2023-03-08 213327" src="https://user-images.githubusercontent.com/69134468/223765551-f8022728-b5c1-4da4-a72f-e578742a8980.png">

Why average is wrong ❌? And how to find the correct sum also?

**The average should be: 3.5**

**Comment by ashleydavis** [comment]
That's strange!

Can you please provide a copy of your code and data so I can try running it and see if I can find the bug.

**Comment by triposat** [comment]
```javascript
// Import modules
const dataForge = require("data-forge");
require("data-forge-fs");

// Path to CSV file
const csvFilePath = "sample_file.csv";

// Load the CSV file into a data frame
const dataFrame = dataForge.readFileSync(csvFilePath).parseCSV();

console.log(dataFrame)

// Group the data by column 'product' and aggregate
const aggregatedData = dataFrame
    .groupBy((row) => row.product)
    .select((group) => ({
        product: group.first().product,
        count: group.count(),
        averagePrice: group.deflate((row) => row.price).average(),
        maxPrice: group.deflate((row) => row.price).max(),
        minPrice: group.deflate((row) => row.price).min(),
    }))
    .inflate();

// Display the aggregated data in the console
console.log(aggregatedData.toArray());
```

**Comment by ashleydavis** [comment]
Data as well pls.

**Comment by triposat** [comment]
date, product, price, quantity
01-03-2023,Apples,3,10
02-03-2023,Bananas,2,20
03-03-2023,Oranges,4,5
04-03-2023,Pineapples,6,15
05-03-2023,Strawberries,8,8
06-03-2023,Blueberries,7,12
09-03-2023,Grapes,9,5
10-03-2023,Kiwis,8,15
13-03-2023,Bananas,3,10
14-03-2023,Oranges,5,5
15-03-2023,Pineapples,7,15
16-03-2023,Strawberries,9,8
17-03-2023,Blueberries,8,12
20-03-2023,Grapes,10,5
21-03-2023,Kiwis,9,15
22-03-2023,Mangoes,12,12
23-03-2023,Papayas,6,20
24-03-2023,Bananas,3,10
25-03-2023,Oranges,5,5
26-03-2023,Pineapples,8,15
30-03-2023,Apples,4,10
31-03-2023,Grapes,10,5
01-04-2023,Kiwis,9,15
02-04-2023,Mangoes,12,12
03-04-2023,Papayas,6,20

**Comment by ashleydavis** [comment]
I think this is because the average function is being run on strings! You haven't parsed you CSV file correctly.

Here's what it looks like for me in Data-Forge Notebook, I've highlighted the code you must add that it shows the correct results:

![image](https://user-images.githubusercontent.com/625126/223919666-c01e5fa7-4ed8-4d10-9d7d-aa15a14e7d1b.png)

**Comment by ashleydavis** [comment]
I've published a new version 1.10.0 that throws an exception if you try to use the maths functions on a series containing strings (or any type other than numbers).

This will ensure no else suffers this confusion.

Thanks so much for your contribution @triposat

**Comment by triposat** [comment]
One more issue - When I am running the below code for the same data, then why are columns so badly formatted, with titles not matching the columns.

```javascript
// Import modules
const dataForge = require("data-forge");
require("data-forge-fs");

// Path to CSV file
const csvFilePath = "sample_file.csv";

// Load the CSV file into a data frame
const dataFrame = dataForge
  .readFileSync(csvFilePath)
  .parseCSV()
  .parseInts("price", "quantity");

// Group the data by column 'product' and count the values
const groupedData = dataFrame.groupBy((row) => row.product);

// Display grouped data
console.log(groupedData.toString());
```

**Columns so badly formatted with titles not matching the columns:**
![tK33FRB](https://user-images.githubusercontent.com/69134468/224889239-bd923238-7c0f-48ab-bc86-002ce00bf100.png)

**Comment by triposat** [comment]
Can anyone help me, please?

**Comment by triposat** [comment]
Can anyone help me, please?

**Comment by ashleydavis** [comment]
If you believe there is a bug please submit and an issue with example code.

If you'd like to contribute please fix the problem and submit a PR.
